### PR TITLE
Update to .NET Standard 2.1 and clean up

### PR DESCRIPTION
**Pull request by wforney**

### CODE REVIEW

**Comment by wforney** [comment]
I also added an EnumerateAsync method using IAsyncEnumerable<TItem> to the repository, among other things. Feel free to pick pieces if you like.

File: .editorconfig [hunk 1]
```diff
@@ -18,16 +18,16 @@ trim_trailing_whitespace = true
...
 
 # this. preferences 
-dotnet_style_qualification_for_field = false:error 
-dotnet_style_qualification_for_property = false:error 
-dotnet_style_qualification_for_method = false:error 
-dotnet_style_qualification_for_event = false:error
+dotnet_style_qualification_for_field = true:error 
+dotnet_style_qualification_for_property = true:error 
+dotnet_style_qualification_for_method = true:error 
+dotnet_style_qualification_for_event = true:error
```
Comment (IEvangelist) [inline .editorconfig:30]:
```suggestion
dotnet_style_qualification_for_field = false:error
dotnet_style_qualification_for_property = false:error
dotnet_style_qualification_for_method = false:error
dotnet_style_qualification_for_event = false:error
```

I do not like the `this` qualifier, please leave these as is - it is a style choice that I prefer and a redundancy that I would not like to introduce.

File: Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs [hunk 2]
```diff
@@ -1,37 +1,36 @@
...
-using Microsoft.Azure.CosmosRepository;
+﻿// Copyright (c) IEvangelist. All rights reserved. Licensed under the MIT License.
 
 namespace ServiceTier
 {
+    using System;
+    using System.Collections.Generic;
+    using System.Linq.Expressions;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.CosmosRepository;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs:9]: I prefer `using` statements to be outside the namespace, not within - please undo these changes as well.

File: Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs [hunk 3]
```diff
@@ -1,37 +1,36 @@
...
+
     public class ExampleService : IExampleService
     {
-        readonly IRepository<Person> _personRepository;
+        private readonly IRepository<Person> personRepository;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs:13]: I prefer to omit default access modifiers, for example `private` on fields - this is again just a style preference.

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosPartitionKeyPathProviderTests.cs [hunk 4]
```diff
@@ -28,6 +27,6 @@ public class PickleChipsItem : Item
         [JsonProperty(PropertyName = "pickles")]
         public string PartitionBits { get; set; } = "Hey, where's the chips?!";
 
-        protected override string GetPartitionKeyValue() => PartitionBits;
+        protected override string GetPartitionKeyValue() => this.PartitionBits;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosPartitionKeyPathProviderTests.cs:30]:
```suggestion
        protected override string GetPartitionKeyValue() => PartitionBits;
```

The `this` qualifier is redundant.

File: Microsoft.Azure.CosmosRepository/src/Providers/ICosmosContainerProvider.cs [hunk 5]
```diff
@@ -1,23 +1,22 @@
...
 namespace Microsoft.Azure.CosmosRepository.Providers
 {
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos;
+    using Microsoft.Azure.CosmosRepository.Options;
+
     /// <summary>
-    /// The cosmos container provider exposes a means of providing
-    /// an instance to the configured <see cref="Container"/> object.
+    /// The cosmos container provider exposes a means of providing an instance to the configured
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/ICosmosContainerProvider.cs:10]: Did you rely on some formatting tool, or something? Curious as I did not...

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosPartitionKeyPathProvider.cs [hunk 6]
```diff
@@ -1,28 +1,28 @@
...
 
+        private static readonly Type partitionKeyNameAttributeType = typeof(PartitionKeyPathAttribute);
+
         /// <inheritdoc />
         public string GetPartitionKeyPath<TItem>() where TItem : Item =>
-            _partionKeyNameMap.GetOrAdd(typeof(TItem), GetPartitionKeyNameFactory);
+            partionKeyNameMap.GetOrAdd(typeof(TItem), GetPartitionKeyNameFactory);
 
-        static string GetPartitionKeyNameFactory(Type type)
+        private static string GetPartitionKeyNameFactory(Type type)
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosPartitionKeyPathProvider.cs:22]:
```suggestion
        static string GetPartitionKeyNameFactory(Type type)
```

Redundant access modifier.

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosPartitionKeyPathProvider.cs [hunk 7]
```diff
@@ -1,28 +1,28 @@
...
-        static readonly ConcurrentDictionary<Type, string> _partionKeyNameMap =
+        private static readonly ConcurrentDictionary<Type, string> partionKeyNameMap =
             new ConcurrentDictionary<Type, string>();
 
+        private static readonly Type partitionKeyNameAttributeType = typeof(PartitionKeyPathAttribute);
+
         /// <inheritdoc />
         public string GetPartitionKeyPath<TItem>() where TItem : Item =>
-            _partionKeyNameMap.GetOrAdd(typeof(TItem), GetPartitionKeyNameFactory);
+            partionKeyNameMap.GetOrAdd(typeof(TItem), GetPartitionKeyNameFactory);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosPartitionKeyPathProvider.cs:20]:
```suggestion
            _partionKeyNameMap.GetOrAdd(typeof(TItem), GetPartitionKeyNameFactory);
```

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosContainerProvider.cs [hunk 8]
```diff
@@ -1,96 +1,101 @@
...
+                            Properties.Resources.TheInsertNameHereIsRequiredWhenContainerPerItemTypeIsFalse,
+                            nameof(DefaultCosmosContainerProvider<TItem>.options.DatabaseId)));
                 }
-                if (_options.ContainerId is null)
+
+                if (this.options.ContainerId is null)
                 {
-                    throw new ArgumentNullException(
-                        $"The {nameof(_options.ContainerId)} is required when container per item type is false.");
+                    throw new NullReferenceException(
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosContainerProvider.cs:61]: I understand the formatting of the string, and the usage of the resource for storing it. Curious about the choice to change the `nameof`, using the `_options.ContainerId` works and is a lot less verbose. Also, you changed the type of the exception here, which is not desirable. I prefer to say that if part of the options are missing, then the argument is `null` as it is part of the parameter that we're validating against.

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosContainerProvider.cs [hunk 9]
```diff
@@ -1,96 +1,101 @@
...
+    /// <inheritdoc />
     internal class DefaultCosmosContainerProvider<TItem>
         : ICosmosContainerProvider<TItem> where TItem : Item
     {
-        readonly Lazy<Task<Container>> _lazyContainer;
-        readonly RepositoryOptions _options;
-        readonly ICosmosClientProvider _cosmosClientProvider;
-        readonly ICosmosPartitionKeyPathProvider _cosmosPartitionKeyPathProvider;
-        readonly ILogger<DefaultCosmosContainerProvider<TItem>> _logger;
+        private readonly ICosmosClientProvider cosmosClientProvider;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosContainerProvider.cs:17]: Again, all redundant access modifiers and `this` qualifiers should be reverted. Name private fields with `_` prefix, please.

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosContainerProvider.cs [hunk 10]
```diff
@@ -1,96 +1,101 @@
...
-using Microsoft.Extensions.Options;
+﻿// Copyright (c) IEvangelist. All rights reserved. Licensed under the MIT License.
 
 namespace Microsoft.Azure.CosmosRepository.Providers
 {
-    /// <inheritdoc/>
+    using System;
+    using System.Threading.Tasks;
+
+    using Microsoft.Azure.Cosmos;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosContainerProvider.cs:8]: Again, `using` should be at start of file, not within namespace please.

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosClientProvider.cs [hunk 11]
```diff
@@ -1,43 +1,82 @@
-﻿// Copyright (c) IEvangelist. All rights reserved.
-// Licensed under the MIT License.
-
-using System;
-using System.Threading.Tasks;
-using Microsoft.Azure.Cosmos;
-using Microsoft.Azure.CosmosRepository.Options;
-using Microsoft.Extensions.Options;
+﻿// Copyright (c) IEvangelist. All rights reserved. Licensed under the MIT License.
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosClientProvider.cs:1]: This entire file can be reverted to its original version, the dispose pattern isn't needed here since all we're doing is cascading a call to dispose. No reason to maintain state of dispose invocations, provide finalizer, or GC calls, etc. - keep it simple.

File: Microsoft.Azure.CosmosRepository/src/Options/RepositoryOptions.cs [hunk 12]
```diff
@@ -1,68 +1,57 @@
-﻿// Copyright (c) IEvangelist. All rights reserved.
-// Licensed under the MIT License.
-
-using Microsoft.Azure.Cosmos;
+﻿// Copyright (c) IEvangelist. All rights reserved. Licensed under the MIT License.
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Options/RepositoryOptions.cs:1]: I'm not sure I like this being formatted this way either, the options were structured in an order that made sense - now it is just alphabetical...which is less ideal for authors.

File: Microsoft.Azure.CosmosRepository/src/Microsoft.Azure.CosmosRepository.csproj [hunk 13]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <Company>Microsoft Corporation</Company>
     <Product>IEvangelist Azure Cosmos Repository</Product>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Microsoft.Azure.CosmosRepository.csproj:6]: Unfortunately, I'd prefer to keep this on .NET Standard 2.0 because it is a much broader spectrum of users. .NET Standard 2.1 is limited with who can consume this. While I fully understand and appreciate the effort, and I too am sad about missing out on some of the features - such as async streams and nullable reference types, I'd like to keep this on 2.0 for now.

**Review by IEvangelist — CHANGES_REQUESTED** [review]
Thank you for taking the time to post this pull request, and spend time trying to improve this project. There are many style changes in this PR that are undesirable, and lots of formatting updates that are making it hard to review. Could you please revert some of the following style changes?

-  Alphabetically ordering members
- `this` qualifier
- Removal of `_` prefix on fields
- Explicit access modifier of defaults, for example `private` could be removed

Also, the other big thing is that I would like to stay on version .NET Standard 2.0 for now as it allows this library to be consumed by a much broader set of runtimes.

File: Microsoft.Azure.CosmosRepository.Samples/ServiceTier/Program.cs [hunk 14]
```diff
@@ -67,8 +66,10 @@ static async Task RawRepositoryExampleAsync(IRepository<Person> repository)
             _ = await repository.CreateAsync(new[] { maryShaw, calvinWeatherfield });
 
             // Reading...
-            Person mary = await repository.GetAsync(maryShaw.Id, maryShaw.SyntheticPartitionKey);
+            Person? mary = await repository.GetAsync(maryShaw.Id, maryShaw.SyntheticPartitionKey);
+            _ = mary ?? throw new NullReferenceException(nameof(mary));
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository.Samples/ServiceTier/Program.cs:70]: Why is this necessary?

File: Microsoft.Azure.CosmosRepository/src/Attributes/PartitionKeyPathAttribute.cs [hunk 15]
```diff
@@ -1,34 +1,31 @@
...
+    /// value will act as the partition key. Partition key paths should start with "/", for example
+    /// "/partition". For more information, see https://docs.microsoft.com/azure/cosmos-db/partitioning-overview.
     /// </summary>
-    /// <remarks>
-    /// By default, "/id" is used.
-    /// </remarks>
+    /// <remarks>By default, "/id" is used.</remarks>
     [AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = true)]
-    public class PartitionKeyPathAttribute : Attribute
+    public sealed class PartitionKeyPathAttribute : Attribute
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Attributes/PartitionKeyPathAttribute.cs:16]: I like this being `sealed` good suggestion.

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 16]
```diff
@@ -1,162 +1,204 @@
...
-            DeleteAsync(id, new PartitionKey(partitionKeyValue ?? id));
-
-        /// <inheritdoc/>
-        public async ValueTask DeleteAsync(string id, PartitionKey partitionKey)
+        /// <inheritdoc />
+        public async ValueTask<TItem> UpdateAsync(TItem value)
         {
-            ItemRequestOptions options = RequestOptions.Options;
-            Container container = await _containerProvider.GetContainerAsync().ConfigureAwait(false);
+            _ = value ?? throw new ArgumentNullException(nameof(value));
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:179]: I'm a bit conflicted about these, but I suppose they do make sense. We should probably add unit tests for these cases.

File: Microsoft.Azure.CosmosRepository/src/Extensions/ParameterRebinder.cs [hunk 17]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) IEvangelist. All rights reserved. Licensed under the MIT License.
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Extensions/ParameterRebinder.cs:1]: I like that this was separated into its own file, kudos.

File: Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs (hunk 2 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs:9]: There was actually a reason for usings being inside the namespace historically. Here is something that explains: https://stackoverflow.com/questions/125319/should-using-directives-be-inside-or-outside-the-namespace#:~:text=Putting%20it%20inside%20the%20namespaces%20makes%20the%20declarations,whether%20they%20go%20outside%20or%20inside%20the%20namespace.

I have been putting using statements inside namespace for a really long time (10+ years) so I'll have to fight my muscle memory on fixing these if you want it that way.

File: .editorconfig (hunk 1 above)
Comment (wforney) [reply .editorconfig:30]: Ok, generally I use the old style guide that was internal to MS a while back and used to be StyleCop oriented. Looks like things have changed a bit since I last updated that. If you use the underscore prefix then you don't need this on everything, but if you don't use the underscore it is making it clear if the variable is local or part of the class.

File: Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs (hunk 3 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository.Samples/ServiceTier/ExampleService.cs:13]: Those were put in by codemaid. I didn't hand enter those so I can flip the switch and revert those easily enough.

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosPartitionKeyPathProviderTests.cs (hunk 4 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosPartitionKeyPathProviderTests.cs:30]: The other reason to use this in these situations is to distinguish between a static and an instance variable. It just prevents you from accidentally doing silly things. I'll go flip the config around a bit and reformat things.

File: Microsoft.Azure.CosmosRepository/src/Providers/ICosmosContainerProvider.cs (hunk 5 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository/src/Providers/ICosmosContainerProvider.cs:10]: Yes, I used the CodeMaid extension and built-in auto-format. It honors editorconfig and whatnot. My main objective was to get it into my standard form, which used to match what Microsoft does internally with their releases. Things have changed there a bit since I put out something there though so I'll have to ping some old friends and see what the newer standards are. I don't think they have guidance on code style across the company any more but I could be wrong.

File: Microsoft.Azure.CosmosRepository/src/Options/RepositoryOptions.cs (hunk 12 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository/src/Options/RepositoryOptions.cs:1]: Alphabetical is the default setting in CodeMaid. I just ran the formatter on it... It does some opinionated stuff... Most of the changes in this PR are cosmetic. There are some actual things though. I'll go make a version that is formatted closer to the original and incrementally change things. That should be a bit easier to read.

File: Microsoft.Azure.CosmosRepository/src/Microsoft.Azure.CosmosRepository.csproj (hunk 13 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository/src/Microsoft.Azure.CosmosRepository.csproj:6]: Ok, fine by me. I'm going to keep my 2.1 version going, with formatting changes to closer match yours. I'll do another PR with just the 2.0 and significant stuff.

File: Microsoft.Azure.CosmosRepository.Samples/ServiceTier/Program.cs (hunk 14 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository.Samples/ServiceTier/Program.cs:70]: GetAsync can return a null so you need the reference nullable Person. The null check after that may just have been me fixing a warning on the next line (not shown), but it would blow up either way without it so I don't suppose it makes a difference.

File: Microsoft.Azure.CosmosRepository/src/Attributes/PartitionKeyPathAttribute.cs (hunk 15 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository/src/Attributes/PartitionKeyPathAttribute.cs:16]: That was a code analysis suggestion from VS. :)

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs (hunk 16 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:179]: Yeah, I put in some of these but I know I didn't catch all of them.

File: Microsoft.Azure.CosmosRepository/src/Extensions/ParameterRebinder.cs (hunk 17 above)
Comment (wforney) [reply Microsoft.Azure.CosmosRepository/src/Extensions/ParameterRebinder.cs:1]: 👍

**Comment by wforney** [comment]
I'll close this one and make another one in a bit.

File: Microsoft.Azure.CosmosRepository/src/Providers/ICosmosContainerProvider.cs (hunk 5 above)
Comment (IEvangelist) [reply Microsoft.Azure.CosmosRepository/src/Providers/ICosmosContainerProvider.cs:10]: I work for Microsoft, the .NET team does have a common *.editorconfig* but each project can have their own thing. I was following what the Azure Cosmos DB .NET SDK repo was doing as it was my hope that they'd eventually adopt this project.
